### ISSUE TITLE
The confusing positional arguments in the definition of focal loss

### ISSUE THREAD

**Issue by LeeJayHui** [qubvel/segmentation_models#258]
Hi, I think you have a small but quite critical mistake in: https://github.com/qubvel/segmentation_models/blob/ba5cfca2b30be14050304096dba2fe0963b588e3/segmentation_models/losses.py#L238
In here, the arguments aboved are passed to the function below in the form of **positional arguments**: https://github.com/qubvel/segmentation_models/blob/ba5cfca2b30be14050304096dba2fe0963b588e3/segmentation_models/base/functional.py#L286
Here comes the confusing part. I think the value of `alpha` which we expect it to be, is replaced with the value of `gamma`. The same thing happens to `gamma`, too. But I'm not 100% certain about this.
If anyone sees this issue, I hope you can give a clear answer.

**Comment by LeeJayHui** [comment]
One more thing, the implementation of binary focal loss here may be slightly different from the formula of widely accepted binary focal loss. 
This issue [https://github.com/qubvel/segmentation_models/issues/187](url) explain this problem.
@qubvel Hope you can help with this.

**Comment by qubvel** [comment]
Hi @LeeJayHui 
Thanks for pointing out about mostake in positional arguments! Can you create a PR to fix it?
About `alpha` coefficient.. It not clear (for me) from paper is it should be `alpha` and `(1 - alpha)` or just `alpha`. Because in paper it is noted as `alpha_t` without description.

**Comment by qubvel** [comment]
I make a fix about positional arguments #259 , but if you would like to contribute, I will merge your PR.

What happens with alpha in case of categorical focal loss if for binary we have alpha and (1 - alpha)?

**Comment by LeeJayHui** [comment]
Thanks for your reply and confirm @qubvel . I check the paper [https://arxiv.org/abs/1708.02002v2](url) and in section **3.1. Balanced Cross Entropy**, the original paper goes as: 

> A common method for addressing class imbalance is to
introduce a weighting factor α ∈ [0,1] for class 1 and 1−α
for class −1. In practice α may be set by inverse class fre-
quency or treated as a hyperparameter to set by cross valida-
tion.

I think it means that in binary task, we should use (1-alpha) for the background class to address class imbalance.
I haven't created a PR yet. I am new to GitHub and Git. I am still trying to figure out how to use it. Thanks a lot to fix this problem. And this repository is really helpful. Good job!

**Comment by LeeJayHui** [comment]
I don't know how to reply your reply, so maybe all I can do is mention you. 
@qubvel  As for multi-class focal loss, I think alpha is kind of confusing here. 
Maybe it is potential to set alpha as the inverse class frequency. For example, assume we have ['dog', 'cat', 'rabbit'] and their corresponding frequencies are [0.1, 0.3, 0.6], we can set the alphas to be [1-0.1, 1-0.3, 1-0.6].
But from my perpesctive, muti-class focal loss is not widely used in CV tasks.

**Comment by qubvel** [comment]
Ok, (1- alpha) sounds good for binary focal loss. Fix proposed in same PR #259 
Will check this one more time and merge to master. You could install __from source__ (see installation instructions in README) to get fixed version.

**Comment by qubvel** [comment]
It would be nice if you check it also and report if something wrong, thank you!

**Comment by LeeJayHui** [comment]
Thanks a lot for your reply and fix! I have already modified the souce code in my local environment. And it did work. Good luck to your future job!
One last thing, I think the decoder parts here can be improved, too. Or we can add more choices of decoder. The plain stack of convolution/bn/act layers in decoder stages may not be the best choice. Any way, thanks again.

**Comment by qubvel** [comment]
You are absolutely right, there are a lot of ways for improvement! If you have any ideas you can open a Pull Request and propose changes!
